zkchannels-crypto/examples/range_proof.rs

```diff
@@ -0,0 +1,160 @@
+    ) {
+        Ok(proof) => proof,
+        Err(e) => {
+            unreachable!("Impossible error while generating proof: {}", e);
+        }
+    };
+
+    // Anyone can verify the proof without knowing your number
+    match random_number_proof.verify(&pedersen_parameters, &range_constraint_parameters) {
+        Ok(()) => println!("Yay! Proof verified!"),
+        Err(e) => eprintln!("Invalid zk proof! {}", e),
+    };
+
+    // You can't make a proof on a number that isn't in the range...
+    match RangeProof::new(
+        &mut rng,
+        &pedersen_parameters,
+        &range_constraint_parameters,
+        -100,
+    ) {
+        Ok(_proof) => unreachable!("Impossible error generating range proof on an invalid value!"),
+        Err(e) => println!("Got expected error: {}", e),
+    };
+
+    // Later, you can reveal the number by sharing it and the blinding factor for the commitment
+    assert!(random_number_proof.open_commitment(&pedersen_parameters, blinding_factor, 12345));
+
+    // You can't pretend to have committed to a different number
+    assert!(!random_number_proof.open_commitment(&pedersen_parameters, blinding_factor, 888));
+}
+
+#[derive(Debug, Error)]
+pub enum RangeError {
+    #[error(transparent)]
+    OutOfRange(#[from] ValueOutsideRange),
+    #[error("Commitment proof is not consistent")]
+    InconsistentCommitment,
+    #[error("Range constraint is not satisfied")]
+    InconsistentRangeConstraint,
+}
+// Zero-knowledge proof of knowledge of a number within the range [0, 2^63).
+pub struct RangeProof {
+    proof: CommitmentProof<G2Projective, 1>,
+    range_constraint: RangeConstraint,
+}
+
+impl RangeProof {
+    /// Try to create a new `RandomNumberProof`.
+    /// This fails if the provided number is out of range.
+    pub fn new(
+        rng: &mut impl Rng,
+        pedersen_parameters: &PedersenParameters<G2Projective, 1>,
+        range_constraint_parameters: &RangeConstraintParameters,
+        number: i64,
+    ) -> Result<(BlindingFactor, Self), RangeError> {
+        // Generate range constraint on the value
+        let range_constraint_builder = RangeConstraintBuilder::generate_constraint_commitments(
+            number,
+            range_constraint_parameters,
+            rng,
+        )?;
+
+        // Commit to the number, using the commitment scalar from the range constraint
+        let proof_builder = CommitmentProofBuilder::generate_proof_commitments(
+            rng,
+            Message::from(Scalar::from(number as u64)),
+            &[Some(range_constraint_builder.commitment_scalar())],
+            pedersen_parameters,
+        );
+
+        let challenge = ChallengeBuilder::new()
+            .with(&range_constraint_builder)
+            .with(&range_constraint_parameters)
+            .with(&proof_builder)
+            .with(pedersen_parameters)
+            .finish();
+
+        // Form the proof
+        Ok((
+            proof_builder.message_blinding_factor(),
```

Comment:
Why is this value included here?

Reply:
It's returned from the function so that the user can later reveal it and open the commitment. They shouldn't send it with the proof to the verifier. I can clarify this in the docs.

Reply:
This seems like a point that might be easy to mess up, yeah.

Reply:
Added clarification